File: ios/StillPointShared/Tests/StillPointSharedTests/SharedFixtures.swift
import Foundation
import XCTest
@testable import StillPointShared

/// Loads the cross-platform golden fixtures in `shared/test-fixtures/` (#421).
/// The same JSON files back the web `vitest` suites via
/// `src/lib/testing/sharedFixtures.ts`, so web + iOS assert against one source of
/// truth and any drift between the duplicated algorithms breaks CI.
///
/// `swift test` compiles from the checked-out source tree, so the fixtures are
/// resolved from this file's `#filePath` (valid locally and in CI).
enum SharedFixtures {
    /// `.../ios/StillPointShared/Tests/StillPointSharedTests/SharedFixtures.swift`
    /// → repo root is five path components up.
    static let directoryURL: URL = {
        var url = URL(fileURLWithPath: #filePath)
        for _ in 0..<5 { url.deleteLastPathComponent() }
        return url.appendingPathComponent("shared/test-fixtures", isDirectory: true)
    }()

    static func load<T: Decodable>(_ fileName: String, as type: T.Type) throws -> T {
        let fileURL = directoryURL.appendingPathComponent(fileName)
        let data = try Data(contentsOf: fileURL)
        return try JSONDecoder().decode(T.self, from: data)
    }
}

// MARK: - Fixture shapes (mirror shared/test-fixtures/*.json)

struct ClearPercentFixture: Decodable {
    struct LogEntry: Decodable {
        let time: Double
        let state: String
    }
    struct Case: Decodable {
        let name: String
        let log: [LogEntry]
        let endTime: Double
        let expected: Int
    }
    let cases: [Case]
}

struct HistoryJourneyFixture: Decodable {
    struct Session: Decodable {
        let id: String
        let sessionType: String
        let sessionDate: String
        let createdAt: String
    }
    /// Flat, discriminated row (`kind`) so heterogeneous rows decode with plain Codable.
    struct ExpectedRow: Decodable {
        let kind: String
        let date: String?
        let startDate: String?
        let endDate: String?
        let dayCount: Int?
        let id: String?
        let sessionIndexInDay: Int?
    }
    struct Case: Decodable {
        let name: String
        let todayIsoDate: String?
        let sessions: [Session]
        let expectedRows: [ExpectedRow]
    }
    let collapseGapThresholdDays: Int
    let cases: [Case]
}

struct SessionStatsFixture: Decodable {
    struct Session: Decodable {
        let id: String
        let sessionType: String
        let dayNumber: Int
        let duration: Int
        let bonusSeconds: Int
        let actualTime: Int
        let completed: Bool
        let clearPercent: Int
        let thoughtCount: Int
        let sessionDate: String
    }
    struct Expected: Decodable {
        let streak: Int
        let avgClearPercent: Int
        let avgThoughtsPerSession: Double
        let avgThoughtsPerMinute: Double
        let bonusSecondsTotal: Int
        let bonusMinutesTotal: Int
    }
    struct Case: Decodable {
        let name: String
        let sessions: [Session]
        let expected: Expected
    }
    let cases: [Case]
}

struct BreathCountingFixture: Decodable {
    struct CountCase: Decodable {
        let taps: Int
        let expected: Int
    }
    struct PhaseCase: Decodable {
        let taps: Int
        let expected: String
    }
    struct DisplayCase: Decodable {
        let seconds: Int
        let expected: String
    }
    let breathCount: [CountCase]
    let phase: [PhaseCase]
    let elapsedDisplay: [DisplayCase]
}

struct AphorismsFixture: Decodable {
    struct Case: Decodable {
        let day: Int
        let expectedIndex: Int
    }
    let expectedCount: Int
    let cases: [Case]
}

struct BuddyCalendarColorsFixture: Decodable {
    struct Case: Decodable {
        let userId: String
        let expectedIndex: Int
        let expectedHex: String
    }
    let paletteLength: Int
    let palette: [String]
    let cases: [Case]
}

struct PathwayFixture: Decodable {
    struct NodeStateCase: Decodable {
        let day: Int
        let currentDay: Int
        let expected: String
    }

    struct ExpectedLevel: Decodable {
        let level: Int
        let name: String
        let state: String
        let completedCount: Int
        let firstNodeState: String?
        let lastNodeState: String?
        let nodeStates: [String]?
    }

    struct BuildPathwayCase: Decodable {
        let name: String
        let currentDay: Int
        let expectedCurrentNodeDay: Int?
        let expectedCurrentNodeCount: Int?
        let expectedLevelCount: Int?
        let expectedAllDays: [Int]?
        let expectedLevels: [ExpectedLevel]?
        let expectedAllNodesCompleted: Bool?
        let expectedAllLevelsCompleted: Bool?
        let expectedLastLevelState: String?
        let expectedLastNodeState: String?
        let expectedFirstNodeState: String?
    }

    let daysPerLevel: Int
    let totalLevels: Int
    let pathwayMaxDay: Int
    let levelNames: [String]
    let nodeStateForDay: [NodeStateCase]
    let buildPathway: [BuildPathwayCase]
}

struct DurationForDayFixture: Decodable {
    struct Constants: Decodable {
        let baseDuration: Int
        let increment: Int
        let maxDuration: Int
        let forkDay: Int
        let recoveryMaxSteps: Int
        let missedDayGapThreshold: Int
    }
    struct DurationCase: Decodable {
        let day: Int
        let expected: Int
    }
    struct EligibilityCase: Decodable {
        let currentDay: Int
        let expected: Bool
    }
    let constants: Constants
    let durationForDay: [DurationCase]
    let isDualTrackEligible: [EligibilityCase]
}

// MARK: - Fixture -> SessionDTO helpers

extension HistoryJourneyFixture.Session {
    func toSessionDTO() -> SessionDTO {
        SessionDTO(
            id: id,
            dayNumber: 1,
            sessionType: SessionType(rawValue: sessionType)!,
            duration: 60,
            completed: true,
            actualTime: 60,
            clearPercent: 50,
            thoughtCount: 0,
            mindStateLog: nil,
            sessionDate: sessionDate,
            createdAt: createdAt,
            buddySessionId: nil
        )
    }
}

extension SessionStatsFixture.Session {
    func toSessionDTO() -> SessionDTO {
        SessionDTO(
            id: id,
            dayNumber: dayNumber,
            sessionType: SessionType(rawValue: sessionType)!,
            duration: duration,
            bonusSeconds: bonusSeconds,
            completed: completed,
            actualTime: actualTime,
            clearPercent: clearPercent,
            thoughtCount: thoughtCount,
            mindStateLog: nil,
            sessionDate: sessionDate,
            createdAt: nil,
            buddySessionId: nil
        )
    }
}

File: src/lib/testing/sharedFixtures.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

/**
 * Loader for the cross-platform golden fixtures in `shared/test-fixtures/`
 * (#421). The same JSON files are consumed by the iOS `swift test` suite via
 * `SharedFixtures.swift`, so web + iOS assert against one source of truth and
 * any drift between the duplicated algorithms breaks CI.
 */
const FIXTURES_DIR = join(
  dirname(fileURLToPath(import.meta.url)),
  "..",
  "..",
  "..",
  "shared",
  "test-fixtures",
);

export function loadSharedFixture<T>(fileName: string): T {
  const raw = readFileSync(join(FIXTURES_DIR, fileName), "utf8");
  return JSON.parse(raw) as T;
}

// MARK: - Fixture shapes (mirror shared/test-fixtures/*.json)

export type ClearPercentFixture = {
  cases: Array<{
    name: string;
    log: Array<{ time: number; state: string }>;
    endTime: number;
    expected: number;
  }>;
};

export type HistoryJourneyFixture = {
  collapseGapThresholdDays: number;
  cases: Array<{
    name: string;
    todayIsoDate: string | null;
    sessions: Array<{
      id: string;
      sessionType: string;
      sessionDate: string;
      createdAt: string;
    }>;
    expectedRows: Array<
      | { kind: "missed"; date: string }
      | { kind: "missedRange"; startDate: string; endDate: string; dayCount: number }
      | { kind: "session"; id: string; sessionIndexInDay: number }
    >;
  }>;
};

export type SessionStatsFixture = {
  cases: Array<{
    name: string;
    sessions: Array<{
      id: string;
      sessionType: string;
      dayNumber: number;
      duration: number;
      bonusSeconds: number;
      actualTime: number;
      completed: boolean;
      clearPercent: number;
      thoughtCount: number;
      sessionDate: string;
    }>;
    expected: {
      streak: number;
      avgClearPercent: number;
      avgThoughtsPerSession: number;
      avgThoughtsPerMinute: number;
      bonusSecondsTotal: number;
      bonusMinutesTotal: number;
    };
  }>;
};

export type BreathCountingFixture = {
  breathCount: Array<{ taps: number; expected: number }>;
  phase: Array<{ taps: number; expected: "inhale" | "exhale" }>;
  elapsedDisplay: Array<{ seconds: number; expected: string }>;
};

export type AphorismsFixture = {
  expectedCount: number;
  cases: Array<{ day: number; expectedIndex: number }>;
};

export type BuddyCalendarColorsFixture = {
  paletteLength: number;
  palette: string[];
  cases: Array<{ userId: string; expectedIndex: number; expectedHex: string }>;
};

export type PathwayFixture = {
  daysPerLevel: number;
  totalLevels: number;
  pathwayMaxDay: number;
  levelNames: string[];
  comingSoonMessage: string;
  nodeStateForDay: Array<{ day: number; currentDay: number; expected: "completed" | "current" | "locked" }>;
  buildPathway: Array<{
    name: string;
    expectedLevelCount?: number;
    expectedAllDays?: number[];
    expectedAllNodesComingSoon?: boolean;
    expectedAllLevelsComingSoon?: boolean;
    expectedAllCompletedCountsZero?: boolean;
    expectedLevels?: Array<{
      level: number;
      name: string;
      state: "completed" | "current" | "locked" | "comingSoon";
      completedCount: number;
      firstNodeState?: "completed" | "current" | "locked" | "comingSoon";
      lastNodeState?: "completed" | "current" | "locked" | "comingSoon";
      nodeStates?: Array<"completed" | "current" | "locked" | "comingSoon">;
    }>;
  }>;
};

export type DurationForDayFixture = {
  constants: {
    baseDuration: number;
    increment: number;
    maxDuration: number;
    forkDay: number;
    recoveryMaxSteps: number;
    missedDayGapThreshold: number;
  };
  durationForDay: Array<{ day: number; expected: number }>;
  isDualTrackEligible: Array<{ currentDay: number; expected: boolean }>;
  advanceProgression: Array<{
    name: string;
    sessionType: "standard" | "quick" | "breath";
    completed: boolean;
    state: {
      currentDay: number;
      recoveryTargetDay: number | null;
      recoveryCurrentStep: number | null;
      recoveryTotalSteps: number | null;
    };
    expected: {
      currentDay: number;
      recoveryTargetDay: number | null;
      recoveryCurrentStep: number | null;
      recoveryTotalSteps: number | null;
    };
  }>;
  detectMissedDayGap: Array<{
    name: string;
    lastCompletedSessionDate: string | null;
    todayIso: string;
    currentDay: number;
    recovery: {
      recoveryTargetDay: number | null;
      recoveryCurrentStep: number | null;
      recoveryTotalSteps: number | null;
    };
    expected: {
      recoveryTargetDay: number;
      recoveryCurrentStep: number;
      recoveryTotalSteps: number;
    } | null;
  }>;
};
